src/Components/Components/src/Routing/RouteEntry.cs

```diff
@@ -25,23 +27,53 @@ public RouteEntry(RouteTemplate template, Type handler, string[] unusedRoutePara
 
         internal void Match(RouteContext context)
         {
-            if (Template.Segments.Length != context.Segments.Length)
+            // If there are no optional segments on the route and the length of the route
+            // and the template do not match, then there is no chance of this matching and
+            // we can bail early.
+            var optionalSegments = Template.Segments.Count(template => template.IsOptional);
+            if (optionalSegments == 0 && Template.Segments.Length != context.Segments.Length)
             {
                 return;
             }
 
             // Parameters will be lazily initialized.
             Dictionary<string, object> parameters = null;
+            var numMatchingSegments = 0;
             for (var i = 0; i < Template.Segments.Length; i++)
             {
                 var segment = Template.Segments[i];
-                var pathSegment = context.Segments[i];
+                
+                // If the template contains more segments than the path, then
+                // break out of this for-loop and continue evaluating the rest
+                // of this method. This will happen in the following cases:
+                //
+                // (1) If we are comparing a literal route with a literal template
+                // and the route is shorter than the template.
+                // (2) If we are comparing a template where the last value is an optional
+                // parameter that the route does not provide.
+                if (i >= context.Segments.Length)
+                {
+                    // If we are under condition (1) above then we can stop evaluating
+                    // matches on the rest of this template.
+                    if (!segment.IsParameter && !segment.IsOptional)
+                    {
+                        break;
+                    }
+                }
+
+                string pathSegment = null;
+                if (i < context.Segments.Length)
```

Comment:
Isn't this always true now (because of the check on Line 54)?

Reply:
I might be mistaken, but if we `pathSegment` is never null at this point, do you still need the `OptionalTypeRouteConstraint` type?

Reply:
It's not always true. We only break in L54 if the segment is not a parameter or an optional. We'll end up hitting this case if the template is something like `/a/{b?}/{c?}` and the route the user provides is `/a/`.

I realize the comment in the break-statement above is a little misleading so I've updated it to reflect that the if statement in L54 is not a guard against out of bounds exceptions.